### PR TITLE
[TRITON] mHC/mHC-lite: Manifold-constrained Hyper Connection

### PR DESCRIPTION
**Pull request by anhminhnguyenhoang**
Co-authors: @waqahmed-amd-fi @anhminhnguyenhoang 
## Motivation

Implements **mHC (manifold-constrained Hyper Connection)** ([arXiv:2512.24880v2](https://arxiv.org/abs/2512.24880v2)) and **mHC-lite** ([arXiv:2601.05732v1](https://arxiv.org/abs/2601.05732v1)) as fused Triton kernels for AMD Instinct GPUs (MI300X / MI355X).

mHC constrains dynamic residual matrices to the Birkhoff polytope via iterative Sinkhorn-Knopp (SK) normalization to prevent gradient explosions in multi-stream residual connections. However, finite SK iterations don't guarantee exact doubly stochasticity and require specialized kernels. mHC-lite addresses this by reparameterizing doubly stochastic matrices as convex combinations of permutation matrices (Birkhoff-von Neumann theorem), achieving **exact** doubly stochasticity in a single softmax + matmul, eliminating the need for iterations.

## Technical Details

**Triton kernels** (`aiter/ops/triton/_triton_kernels/fusions/mhc.py`):
- `_mhc_fused_kernel` — fused RMSNorm, 3-stream projection matmul, scaling + bias, activations (H_pre and H_post for mHC and H_pre, H_res, H_post for mHC-lite)
- `_mhc_fused_split_kernel` / `_mhc_fused_reduce_kernel` — split-K variant for large K dimensions
- `_sinkhorn_knopp_log_domain_kernel` — log-domain SK for H_res in mHC

**Two H_res modes for _mhc_fused_kernel**:
- **Sinkhorn mode** (`mode="mhc"`): Standard mHC per [arXiv:2512.24880v2](https://arxiv.org/abs/2512.24880v2) — the fused kernel outputs raw logits for H_res (shape M × n²), followed by iterative Sinkhorn-Knopp (default 20 iterations) to produce an approximate doubly stochastic matrix. i.e. raw logits → iterative SK.
- **Lite mode** (`mode="mhc_lite"`): mHC-lite per [arXiv:2601.05732v1](https://arxiv.org/abs/2601.05732v1) — `phi_res` projects to n! outputs, the kernel applies softmax to obtain convex coefficients `alpha`, then computes `H_res = alpha @ P` where P is all n! permutation matrices via a single matrix multiply against all n! permutation matrices. This produces an **exact** doubly stochastic matrix in one pass, bypassing SK iterations entirely. For n=4 (the recommended setting from the HC paper), this involves 24 permutation matrices. i.e. `phi_res` projects to n! outputs → softmax → weighted sum over all n! permutation matrices.

**Python API** (`aiter/ops/triton/fusions/mhc.py`): `mhc()`, `sinkhorn_knopp()`

**Config system** (`aiter/ops/triton/utils/mhc_config_utils.py`): arch-based config loading with M and C dim threshold matching and tuned JSON configs for gfx942 (MI300X) and gfx950 (MI355X).

**Test utilities** (`op_tests/triton_tests/utils/mhc_ref.py`): PyTorch reference implementations for both modes + input generators.

**Benchmarks** (`op_tests/op_benchmarks/triton/bench_mhc.py`): perf reporting across M, C, n dimensions.

**Further documentation:** https://amd.atlassian.net/wiki/spaces/AIG/pages/1388423373/Triton+Kernels+mHC

## Test Plan

Pytest suite (`op_tests/triton_tests/fusions/test_mhc.py`) covering:
- Correctness vs PyTorch reference for both Sinkhorn and lite modes (bf16/fp16)
- Split-K correctness
- Standalone Sinkhorn-Knopp correctness and convergence
- Lite mode exact doubly stochastic verification
- Output range validation and edge cases (zero input, large values, epsilon/alpha variations)

### Launch docker container
Replace 'your_NTID' as needed
```
docker run -it --rm --name mhc-dev --device=/dev/kfd --device=/dev/dri \
  --group-add video --ipc=host --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
  -v /home/your_NTID/aiter:/workspace/aiter -w /workspace/aiter \
  rocm/pytorch:rocm7.1_ubuntu22.04_py3.10_pytorch_release_2.9.1 \
  bash

# inside docker
pip install -e .
```

### Run tests
Pytest suite
```
pytest op_tests/triton_tests/fusions/test_mhc.py -v -s
```
Full benchmark
```
bash op_tests/triton_tests/fusions/bench_mhc.sh
```
Single mode test
```
python op_tests/op_benchmarks/triton/bench_mhc.py --mode mhc
python op_tests/op_benchmarks/triton/bench_mhc.py --mode mhc_lite
python op_tests/op_benchmarks/triton/bench_mhc.py --mode sinkhorn_knopp_only
```

## Test Result
- Tolerances: 1e-2 for small C, 5e-2 for large C (bf16 accumulation).
- gfx942 benchmarks:
```
# mhc
          M    n        C  time(ms)  throughput(TFLOPS)  bandwidth(GB/s)  arithmetic_intensity(FLOP/byte)
0    1024.0  4.0    128.0  0.062026            0.417146        33.650480                        12.551693
1    2048.0  4.0    128.0  0.072431            0.843930        67.123288                        12.615655
2    4096.0  4.0    128.0  0.073465            1.687988       133.384624                        12.647881
3    8192.0  4.0    128.0  0.095721            2.563300       202.604290                        12.664056
4   16384.0  4.0    128.0  0.155839            3.145990       248.335782                        12.672159
5    1024.0  4.0    512.0  0.073062            1.555297        77.817717                        19.872126
6    2048.0  4.0    512.0  0.072856            3.088577       155.599128                        20.046612
7    4096.0  4.0    512.0  0.170759            2.624699       131.477908                        20.135009
8    8192.0  4.0    512.0  0.171910            5.241289       261.912643                        20.179500
9   16384.0  4.0    512.0  0.192041            9.413185       468.069581                        20.201820
10   1024.0  4.0   4096.0  0.252424            3.443689       139.388625                        24.528617
11   2048.0  4.0   4096.0  0.250319            7.138402       287.652125                        24.801822
12   4096.0  4.0   4096.0  0.259290           13.529074       546.308855                        24.940720
13   8192.0  4.0   4096.0  0.265442           26.346870      1049.870779                        25.010754
14  16384.0  4.0   4096.0  0.462807           29.125889      1159.836698                        25.045919
15   1024.0  4.0  32768.0  0.343675           20.313505       802.938254                        25.291309
16   2048.0  4.0  32768.0  0.451650           29.771143      1158.537959                        25.582804
17   4096.0  4.0  32768.0  0.741270           37.312587      1453.993985                        25.731085
18   8192.0  4.0  32768.0  1.627093           34.016453      1313.551091                        25.805872
19  16384.0  4.0  32768.0  3.087023           35.686939      1382.942917                        25.843429

#mhc-lite
          M    n        C  time(ms)  throughput(TFLOPS)  bandwidth(GB/s)  arithmetic_intensity(FLOP/byte)
0    1024.0  4.0    128.0  0.033552            1.151895        34.926068                        32.423577
1    2048.0  4.0    128.0  0.033468            2.166965        66.922445                        32.913236
2    4096.0  4.0    128.0  0.032679            4.498294       135.354278                        33.163653
3    8192.0  4.0    128.0  0.040410            7.747147       237.327775                        33.290296
4   16384.0  4.0    128.0  0.050472           11.592599       346.435661                        33.353981
5    1024.0  4.0    512.0  0.035219            4.071225       124.755314                        32.828482
6    2048.0  4.0    512.0  0.038292            7.793442       238.740375                        33.331135
7    4096.0  4.0    512.0  0.044213           13.028275       391.418419                        33.588279
8    8192.0  4.0    512.0  0.044868           26.223534       780.265727                        33.718344
9   16384.0  4.0    512.0  0.068398           33.975556      1025.281466                        33.783755
10   1024.0  4.0   4096.0  0.078053           15.343746       468.943470                        32.951867
11   2048.0  4.0   4096.0  0.089416           25.903047       776.386630                        33.458514
12   4096.0  4.0   4096.0  0.127875           35.722534      1056.496336                        33.717726
13   8192.0  4.0   4096.0  0.218429           39.015255      1156.354241                        33.848843
14  16384.0  4.0   4096.0  0.338392           53.068652      1587.734284                        33.914785
15   1024.0  4.0  32768.0  0.273780           33.283701      1006.778272                        32.967465
16   2048.0  4.0  32768.0  0.400342           44.014215      1352.349712                        33.474619
17   4096.0  4.0  32768.0  0.803122           44.987162      1328.466146                        33.734093
18   8192.0  4.0  32768.0  1.786622           40.572559      1189.636316                        33.865344
19  16384.0  4.0  32768.0  3.917525           36.999570      1088.519937                        33.931353
```

- gfx950 benchmarks
```
# mhc
          M    n        C  time(ms)  throughput(TFLOPS)  bandwidth(GB/s)  arithmetic_intensity(FLOP/byte)
0    1024.0  4.0    128.0  0.039175            0.784103        63.102748                        12.551693
1    2048.0  4.0    128.0  0.043387            1.404580       111.999173                        12.615655
2    4096.0  4.0    128.0  0.050287            2.454369       194.528954                        12.647881
3    8192.0  4.0    128.0  0.074445            3.297888       259.628104                        12.664056
4   16384.0  4.0    128.0  0.125373            3.887457       307.391543                        12.672159
5    1024.0  4.0    512.0  0.051300            2.193968       110.437982                        19.872126
6    2048.0  4.0    512.0  0.056497            4.061660       206.268363                        20.046612
7    4096.0  4.0    512.0  0.064253            7.245049       363.776349                        20.135009
8    8192.0  4.0    512.0  0.090785            9.993113       497.668685                        20.179500
9   16384.0  4.0    512.0  0.142144           12.819367       641.174265                        20.201820
10   1024.0  4.0   4096.0  0.133059            6.720047       278.763628                        24.528617
11   2048.0  4.0   4096.0  0.104438           17.175390       697.964500                        24.801822
12   4096.0  4.0   4096.0  0.113785           30.823678      1235.885409                        24.940720
13   8192.0  4.0   4096.0  0.196818           35.567806      1422.078308                        25.010754
14  16384.0  4.0   4096.0  0.317049           44.180145      1763.186365                        25.045919
15   1024.0  4.0  32768.0  0.288033           24.220523       957.925687                        25.291309
16   2048.0  4.0  32768.0  0.319367           43.766069      1709.273441                        25.582804
17   4096.0  4.0  32768.0  0.493714           56.603592      2199.601414                        25.731085
18   8192.0  4.0  32768.0  0.988865           56.597273      2190.586357                        25.805872
19  16384.0  4.0  32768.0  1.573129           71.185429      2750.751151                        25.843429

# mhc-lite
          M    n        C  time(ms)  throughput(TFLOPS)  bandwidth(GB/s)  arithmetic_intensity(FLOP/byte)
0    1024.0  4.0    128.0  0.018255            2.015630        62.315842                        32.423577
1    2048.0  4.0    128.0  0.018507            3.955632       121.141808                        32.913236
2    4096.0  4.0    128.0  0.023493            6.415517       198.820521                        33.163653
3    8192.0  4.0    128.0  0.017736           17.192982       526.527291                        33.290296
4   16384.0  4.0    128.0  0.020259           27.158355       889.448946                        33.353981
5    1024.0  4.0    512.0  0.018007            8.132284       248.730148                        32.828482
6    2048.0  4.0    512.0  0.021541           14.115307       420.777926                        33.331135
7    4096.0  4.0    512.0  0.025343           23.675708       742.306457                        33.588279
8    8192.0  4.0    512.0  0.034345           35.313457      1072.537235                        33.718344
9   16384.0  4.0    512.0  0.043761           57.245420      1720.406460                        33.783755
10   1024.0  4.0   4096.0  0.041804           27.358050       831.931203                        32.951867
11   2048.0  4.0   4096.0  0.051196           45.123366      1367.042473                        33.458514
12   4096.0  4.0   4096.0  0.082152           55.678254      1652.046046                        33.717726
13   8192.0  4.0   4096.0  0.125201           73.103298      2157.138456                        33.848843
14  16384.0  4.0   4096.0  0.199910           91.526628      2698.565764                        33.914785
15   1024.0  4.0  32768.0  0.190575           48.273754      1463.286895                        32.967465
16   2048.0  4.0  32768.0  0.258559           71.052049      2124.355877                        33.474619
17   4096.0  4.0  32768.0  0.398922           91.536961      2713.145902                        33.734093
18   8192.0  4.0  32768.0  0.758463           96.651660      2852.170178                        33.865344
19  16384.0  4.0  32768.0  1.608276           90.587212      2672.001665                        33.931353
```
## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: op_tests/triton_tests/fusions/test_mhc.py [hunk 1]
```diff
@@ -68,11 +68,13 @@ def test_mhc_correctness(M, n, C, dtype):
...
         rtol=1e-2,
     )
+    # Relaxed tolerance for H_res due to Sinkhorn-Knopp iterative algorithm
+    # which amplifies small numerical differences, especially with bfloat16
     torch.testing.assert_close(
         H_res_triton.to(torch.float32),
         H_res_torch.to(torch.float32),
-        atol=1e-2,
-        rtol=1e-2,
+        atol=5e-2,
```
Comment (anhminhnguyenhoang) [inline op_tests/triton_tests/fusions/test_mhc.py]: Did you run into test failure because of this for similar tests that you need to relax the tolerance?

Comment (waqahmed-amd-fi) [reply op_tests/triton_tests/fusions/test_mhc.py]: Yes, mainly because of sinkhorn which is an iterative process and returns higher differences due you only 10 iterations. May be we can try 20 for better results?

**Review by juuso-oskari — APPROVED** [review]
LGTM
